File: oapp/services/menu/internal/models/product.go
package models

import "time"

// Product represents a menu product
type Product struct {
	ProductID   string   `json:"product_id" db:"product_id"`
	LocationID  string   `json:"location_id" db:"location_id"`
	Name        string   `json:"name" db:"name"`
	Description *string  `json:"description,omitempty" db:"description"`
	Price       float64  `json:"price" db:"price"`
	Category    *string  `json:"category,omitempty" db:"category"`
	ImageURL    *string  `json:"image_url,omitempty" db:"image_url"`
	IsAvailable bool     `json:"is_available" db:"is_available"`
	CreatedAt   time.Time `json:"created_at" db:"created_at"`
	UpdatedAt   time.Time `json:"updated_at" db:"updated_at"`
}

// ProductWithAttributes represents a product with its attribute groups
type ProductWithAttributes struct {
	Product
	AttributeGroupIDs []string `json:"attribute_group_ids,omitempty"`
}

// CreateProductRequest represents the request to create a product
type CreateProductRequest struct {
	Name              string   `json:"name" validate:"required,max=255"`
	Description       *string  `json:"description,omitempty"`
	Price             float64  `json:"price" validate:"required,gt=0"`
	Category          *string  `json:"category,omitempty" validate:"omitempty,max=100"`
	ImageURL          *string  `json:"image_url,omitempty"`
	AttributeGroupIDs []string `json:"attribute_group_ids,omitempty"`
}

// UpdateProductRequest represents the request to update a product
type UpdateProductRequest struct {
	Name              *string  `json:"name,omitempty" validate:"omitempty,max=255"`
	Description       *string  `json:"description,omitempty"`
	Price             *float64 `json:"price,omitempty" validate:"omitempty,gt=0"`
	Category          *string  `json:"category,omitempty" validate:"omitempty,max=100"`
	ImageURL          *string  `json:"image_url,omitempty"`
	AttributeGroupIDs []string `json:"attribute_group_ids,omitempty"`
}

// ToggleAvailabilityRequest represents the request to toggle product availability
type ToggleAvailabilityRequest struct {
	IsAvailable bool `json:"is_available"`
}

File: oapp/DeliveryXOApp/DeliveryXOApp/Models/Product.swift
//
//  Product.swift
//  DeliveryXOApp
//
//  Created on 2025-10-21.
//

import Foundation

/// Represents a product in the store's catalog
struct Product: Codable, Identifiable {
    let id: String
    let storeId: String
    let name: String
    let description: String?
    let basePrice: Double
    let imageUrl: String?
    let isActive: Bool
    let createdAt: Date
    let updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case storeId = "store_id"
        case name
        case description
        case basePrice = "base_price"
        case imageUrl = "image_url"
        case isActive = "is_active"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

/// Represents a product as it appears in a menu with collection info
struct ProductInMenu: Codable, Identifiable {
    let id: String
    let storeId: String
    let name: String
    let description: String?
    let basePrice: Double
    let imageUrl: String?
    let isActive: Bool
    let createdAt: Date
    let updatedAt: Date
    let menuProductId: String
    let collectionId: String
    let displayOrder: Int
    let isAvailable: Bool
    let menuProductCreatedAt: Date
    let menuProductUpdatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case storeId = "store_id"
        case name
        case description
        case basePrice = "base_price"
        case imageUrl = "image_url"
        case isActive = "is_active"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case menuProductId = "menu_product_id"
        case collectionId = "collection_id"
        case displayOrder = "display_order"
        case isAvailable = "is_available"
        case menuProductCreatedAt = "menu_product_created_at"
        case menuProductUpdatedAt = "menu_product_updated_at"
    }
}

/// Request model for creating a new product
struct CreateProductRequest: Codable {
    let name: String
    let description: String?
    let basePrice: Double
    let imageUrl: String?
    let isActive: Bool?

    enum CodingKeys: String, CodingKey {
        case name
        case description
        case basePrice = "base_price"
        case imageUrl = "image_url"
        case isActive = "is_active"
    }
}

/// Request model for adding a product to a menu collection
struct AddProductToMenuRequest: Codable {
    let productId: String
    let displayOrder: Int?
    let isAvailable: Bool?

    enum CodingKeys: String, CodingKey {
        case productId = "product_id"
        case displayOrder = "display_order"
        case isAvailable = "is_available"
    }
}

/// Request model for toggling product availability in menu
struct ToggleAvailabilityRequest: Codable {
    let isAvailable: Bool

    enum CodingKeys: String, CodingKey {
        case isAvailable = "is_available"
    }
}
